Modify rule S2589: Extend LaYC content for Python (#3173)
SonarSource/rspec: rules/S2589/python/rule.adoc

Gratuitous boolean expressions are conditions that do not change the evaluation
of a program.
This issue can indicate logical errors and affect the correctness of an
application, as well as its maintainability.

== Why is this an issue?

Control flow constructs like `if`-statements allow the programmer to direct the
flow of a program depending on a boolean expression.

> However, suppose that condition is always true or always false,
> independent of any input or other circumstances.
> In that case, the program will effectively always execute only one of the
> available execution branches.
> Thus, the control flow construct and the condition no longer serve a purpose,
> they become _gratuitous_.

Reviewer: I think the wording could be a bit lighter here.

```suggestion
However, if the condition is always true or always false, only one of the branches
will be executed.
In that case, the control flow construct and the condition no longer serve a purpose;
they become _gratuitous_.
```

logical errors.
If a mistake was made, it can be corrected so that the condition is no longer
gratuitous.

If it becomes apparent that the condition is actually unnecessary,
it can be removed together with the associated control flow construct
(e.g. the `if`-statement containing the condition).

=== Code examples

==== Noncompliant code example

[source,python,diff-id=1,diff-type=compliant]
----
def f(b):
    a = True
    if a:  # Noncompliant
        do_something()

    if a and b:  # Noncompliant; "a" is always "True"
        do_something_else()
----

==== Compliant solution

[source,python,diff-id=1,diff-type=noncompliant]
----
def f(b):
    a = True
    do_something()

    if b:
        do_something_else()
----

==== Noncompliant code example

[source,python,diff-id=2,diff-type=compliant]
----
def f(a, b):
    if a is None and b is None:
        do_something()
    elif a is not None or b is not None: # Noncompliant
        do_something_else()
----

==== Compliant solution

[source,python,diff-id=2,diff-type=noncompliant]
----
def f(a, b):
    if a is None and b is None:
        do_something()
    else:
        do_something_else()
----

=== How does this work?

> In the first example, the gratuitous condition `a` is a `True` constant.

Reviewer: Nitpick:

```suggestion
In the first example, the gratuitous condition `a` is simply`True`.
```
